### PR TITLE
Improve tabulate

### PR DESCRIPTION
**Pull request by cgarciae**
# What does this PR do?
Adds various small improvements to `tabulate` including:
* Removes `output_methods` argument, it is no longer necessary since now internally `capture_intermediates` is set to a function that collects them automatically (inspired by [this comment](https://github.com/google/flax/pull/1844#issuecomment-1140321111), thanks @trvs2cool!). Also solves the issue of intermediates for methods other than `__call__` not being captured.
* Adds the `exclude_methods` argument which lets you exclude certain methods, `setup` is excluded by default.
* Adds a test for modules with methods
* Sets `summary.tabulate`'s `mutable` argument to `True` by default to be consistent with `Module.tabulate`'s signature.
* Fixes `None`s being rendered as `null`.
* Removes argument default values from `_get_module_table` to avoid potential errors.

### CODE REVIEW

File: tests/linen/summary_test.py [hunk 1]
```diff
@@ -33,38 +33,65 @@ def _maybe_get_shape(x):
   return x.shape if hasattr(x, "shape") else x
 
 
-def _get_tabulate_cnn(test_sow: bool = False) -> Type[nn.Module]:
+def _get_tabulate_cnn(test_sow: bool = False, use_method: bool = False) -> Type[nn.Module]:
   @dataclasses.dataclass
   class ConvBlock(nn.Module):
```
Comment (marcvanzee) [inline tests/linen/summary_test.py:38]: Can't you just have one class with both `__call__` and `block_method` implemented? (maybe with different behavior)

Then you can just make `test_sow` a class attribute as well and you don't need `_get_tabulate_cnn` anymore.

File: tests/linen/summary_test.py [hunk 2]
```diff
@@ -179,8 +207,8 @@ def test_module_summary_with_depth(self):
 
     table = _get_module_table(
       module, 
-      {"dropout":random.PRNGKey(0), "params": random.PRNGKey(1)}, 
-      depth=1,
+      {"dropout":random.PRNGKey(0), "params": random.PRNGKey(1)},
+      method=CNN.__call__, mutable=True, depth=1,
```
Comment (marcvanzee) [inline tests/linen/summary_test.py:211]: Why did you add the `method` kwarg here? It should be the default right?

**Review by marcvanzee — CHANGES_REQUESTED** [review]
One thought about `_maybe_get_shape`: do you really need it? It seems the outputs of the `table[i].outputs` is always either an array or a pair `(array, dict)`. Can't you just compare them directly? E.g. instead of:

```py
self.assertEqual(
      jax.tree_map(_maybe_get_shape, table[0].outputs),
      (x.shape, dict(training=True)),
    )
```

You can do:


```py
self.assertEqual(table[0].outputs, 
      (x.shape, dict(training=True)),
```

Or am I missing something?

File: tests/linen/summary_test.py (hunk 1 above)
Comment (cgarciae) [reply tests/linen/summary_test.py:38]: Sounds reasonable. Will refactor.

File: tests/linen/summary_test.py (hunk 2 above)
Comment (cgarciae) [reply tests/linen/summary_test.py:211]: I removed all defaults from this private `_get_module_table` function, however, `method` here could be set to `None` so that `__call__` is internally used by default.0

**Comment by cgarciae** [comment]
@marcvanzee you'd have to do it like this:

```python
self.assertEqual(
  (table[0].outputs[0].shape, table[0].outputs[1]),
  (x.shape, dict(training=True)
)
```
Doesn't look as bad as I thought. Will refactor.

**Comment by marcvanzee** [comment]
> @marcvanzee you'd have to do it like this:
> 
> ```python
> self.assertEqual(
>   (table[0].outputs[0].shape, table[0].outputs[1]),
>   (x.shape, dict(training=True)
> )
> ```
> 
> Doesn't look as bad as I thought. Will refactor.

Or just as two separate asserts?

```py
self.assertEqual(table[0].outputs[0].shape, x.shape)
self.assertEqual(table[0].outputs[1], dict(training=True))
```

**Comment by cgarciae** [comment]
@marcvanzee refactored `test_summary.py` based on your comments.
Also added the `excluded_methods` argument which lets you exclude certain method, `setup` is excluded by default.

**Review by marcvanzee — APPROVED** [review]
Looks good!
